gobblin-tunnel/src/main/java/gobblin/tunnel/ProxySetupHandler.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (C) 2014-2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.tunnel;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.nio.ByteBuffer;
+import java.nio.channels.SelectionKey;
+import java.nio.channels.Selector;
+import java.nio.channels.SocketChannel;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.Callable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Handler for setting up the connection from the Tunnel to the intermediate proxy via HTTP CONNECT.
+ */
+class ProxySetupHandler implements Callable<HandlerState> {
+  private static final Logger LOG = LoggerFactory.getLogger(Tunnel.class);
+
+  private static final ByteBuffer OK_REPLY = ByteBuffer.wrap("HTTP/1.1 200".getBytes());
+  private static final Set<ByteBuffer> OK_REPLIES = new HashSet<ByteBuffer>(
+      Arrays.asList(OK_REPLY, ByteBuffer.wrap("HTTP/1.0 200".getBytes())));
+
+
+  private final SocketChannel client;
+  private final Selector selector;
+  private final SocketChannel proxy;
+  private HandlerState state = HandlerState.CONNECTING;
+
+  private ByteBuffer buffer;
+  private final long connectStartTime;
+  private int totalBytesRead = 0;
+  private final Config config;
+
+  ProxySetupHandler(SocketChannel client, Selector selector, Config config)
+      throws IOException {
+    this.config = config;
+    this.client = client;
+    this.selector = selector;
+    buffer = ByteBuffer.wrap(
+        String.format("CONNECT %s:%d HTTP/1.1\r\nUser-Agent: GobblinTunnel\r\nservice-name: gobblin\r\n" +
```

Comment:
Can we not use a standard Apache `HttpClient` to do the HTTP CONNECT + to handle the response?